src/udev/udev-node.c

```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
                                 r = -errno;
                 } while (r == -ENOENT);
 
-        return r;
+        for (i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1 = {}, st2 = {};
```

Comment:
no need to initialize

Reply:
Sure, but...

Honestly I disagree. In C programs we tend to get bitten by uninitialised variables from time to time, not initialising also makes back-porting more error prone. Not that this matter in this specific case but in general I don't think leaving variables uninitialised is a good practise (not sure why we do this in systemd).

Moreover, in many cases we opt-in for "less clever" and sometimes slower code just to keep the source code more readable and better maintainable. Hence I don't understand why should we do such micro-optimisations when in comes to variable initialisation.

Reply:
Consistency is paramount in such cases. If `stat` is borked and doesn't initialize the output parameter correctly, then all calls to `stat` need to pre-initialize. We have hundreds of `stat` calls, and it generally is never initialized. The same pattern should be used here.

Reply:
Actually, initialisation is needed in this case. This is because the same code handles removal of symlinks as well. Basically the link directory can get removed before we do second `stat()` and that could lead to comparison of `st1` and `st2` where `st2` contains garbage.

Reply:
Btw, this is not theoretical issue. If I remove it, `udev-test.pl` starts failing immediately.

Reply:
~~OK, failure to initialize `st1` is ignored in certain cases, so preinitialization is required in this case.~~

I take that back: if we fail to initialize `st1` and it remains preinitialized to 0, we know that `stat_inode_unmodified` will always return false. This means that we might just as well not do the second stat. And stat is somewhat expensive, and we shouldn't do it pointlessly. So this needs to be reworked somehow.

Reply:
On addition first `stat()` always succeeds because we create the directory just before entering the loop. On removal, either again first `stat()` either succeeds and we attempt to do also second one (but in parallel directory might get nuked) or it fails and because of `-ENOENT` return from `link_find_prioritized()` we don't even attempt the second.

I can remove initialisation of the first one because that is safe.